src/hotspot/cpu/riscv/macroAssembler_riscv.cpp

```diff
@@ -1167,6 +1167,19 @@ void MacroAssembler::fsflagsi(Register Rd, unsigned imm) {
 
 #undef INSN
 
+void MacroAssembler::restore_cpu_control_state_after_jni(Register tmp) {
+  if (RestoreMXCSROnJNICalls) {
+    Label skip_fsrmi;
+    frrm(tmp);
+    // Set FRM to the state we need. We do want Round to Nearest. We
+    // don't want non-IEEE rounding modes.
+    guarantee(RoundingMode::rne == 0, "must be");
+    beq(tmp, zr, skip_fsrmi);        // Only reset FRM if it's wrong
```

Comment:
is it really better (performance wise) than doing it always, unconditionaly (so minus frrm, minus beq) ?

Reply:
It's done by following reasons:
1. by the optimization guide, https://riscv-optimization-guide.riseproject.dev/#_controlling_rounding_behavior_scalar.
2. aarch64 apply the similar optimization.

Please also check discussion at: https://github.com/openjdk/jdk/pull/18758

Reply:
Thanks for the links, however there are no performance claims in that discussion, only few "maybes" for "some hardware", can we check on existing h/w ( c910, u74) ? with jmh test doing dummy jni calls ?

Reply:
Thanks for the suggesion.
c910 might be suitable for this kind test (u74 is in order, so not?), but I don't have the hardware now.
If anyone happens to have it and can help test it that would be helpful.

Reply:
@luhenry I've already saw a big criticism of that document. So it makes me take it with a grain of salt. It was written by a limited group of vendors ( not all of them participated) and basically shows  u-arch peculiarities of limited hw group.

I'm gonna check few version of this patch in jni_blank jmh microtest

Reply:
@VladimirKempik again, happy to get input from more people on that document! :)

Reply:
Here are results. using:
a) jni_blank as is
b) modified jni_blank where native func() does this:

```
int x = 2;
    asm
        ("csrw fcsr,%0\n\t"
        :
        : "r" (x)
        );
```
branchful - exactly this PR
branchless - this PR without csrr&beq

results:

no fcsr change in native code

```
branchless
Benchmark                       Mode  Cnt    Score   Error  Units
CallOverheadConstant.jni_blank  avgt   30  133.586 ? 1.431  ns/op
CallOverheadVirtual.jni_blank   avgt   30  131.715 ? 0.570  ns/op


branchful
Benchmark                       Mode  Cnt    Score   Error  Units
CallOverheadConstant.jni_blank  avgt   30  133.376 ? 1.491  ns/op
CallOverheadVirtual.jni_blank   avgt   30  133.560 ? 1.782  ns/op
```

fcsr changed to rdn in native code

```

branchless
Benchmark                       Mode  Cnt    Score   Error  Units
CallOverheadConstant.jni_blank  avgt   30  153.708 ? 1.191  ns/op
CallOverheadVirtual.jni_blank   avgt   30  150.653 ? 1.617  ns/op

branchful
Benchmark                       Mode  Cnt    Score    Error  Units
CallOverheadConstant.jni_blank  avgt   30  153.595 ?  0.759  ns/op
CallOverheadVirtual.jni_blank   avgt   30  149.992 ?  1.605  ns/op
```

Basically there are not difference here, so why would you make code more complex (and require additional registers) ?

Reply:
Great to see there is no performance degradation on the current generation of hardware. That's a great motivator to get this change in with the branch as we already know that the branch is going to bring a performance improvement on next/other generations of hardware.

Reply:
Even when showing middle finger to branch predictor:
```
unsigned char prbarray[512];
static int setupme()
{
    srand(time(NULL));
    for(int i = 0; i < 512; i++)
    {
	int r = rand()%15;
	prbarray[i]= (r > 10) ? (r-10) : 0;
    }

    return 1;
}

EXPORT void func() {
    static int setup = 0;
    static unsigned long counter = 0;
    if(!setup) setup = setupme();
    int x = prbarray[counter++%512];
    asm
        ("csrw fcsr,%0\n\t"
        :
        : "r" (x)
        );
}
```

results are still equal between branchless and branchful versions

I'm ok with this change, buts lets wait for yet another opinion: do we accept such "future proof" changes or not

Reply:
I don't have a strong opinion on this. This code will only be enabled by the user on command line in case we are calling some buggy external libraries which may corrupt the FP control register. It won't make a difference for the most normal cases. My local tests show that `frrm` is 3x faster than `fsrmi` on sifive/u74. So it does make sense to have the branch check on platforms like this. It should not be a big issue to other more advanced platforms with a good BranchPredictor.

Reply:
> Suggesion: beqz(tmp, skip_fsrmi);        // Only reset FRM if it's wrong

fixed, thanks!

Reply:
Thanks everyone for testing and discussion!